nix/build-maven-repo.nix

```diff
@@ -0,0 +1,189 @@
+  #        url = "https://repo.maven.apache.org/maven2/com/squareup/okio/okio/3.9.0/okio-3.9.0.jar";
+  #        hash = "...";
+  #        downloadToTemp = true;
+  #        postFetch = "install -Dt $out/com/squareup/okio/okio/3.9.0/ $downloadedFile"
+  #      };
+  #    }
+  #
+  # 2. Patch a JAR containing native binaries:
+  #
+  #    {
+  #      "com.android.tools.build:aapt2:8.5.0-rc02-11315950" = {
+  #        "aapt2-8.5.0-rc02-11315950-linux.jar" = src: runCommandCC src.name {
+  #          nativeBuildInputs = [ jdk autoPatchelfHook ];
+  #          dontAutoPatchelf = true;
+  #        } ''
+  #          cp ${src} aapt2.jar
+  #          jar xf aapt2.jar aapt2
+  #          chmod +x aapt2
+  #          autoPatchelf aapt2
+  #          jar uf aapt2.jar aapt2
+  #          cp aapt2.jar $out
+  #        '';
+  #      }
+  #    }
+  overrides ? { },
+}:
+
+let
+
+  inherit (builtins)
+    attrValues
+    elemAt
+    filter
+    fromJSON
+    getAttr
+    head
+    length
+    mapAttrs
+    replaceStrings
+    ;
+
+  inherit (lib) mapAttrsToList readFile;
+
+  inherit (lib.strings) sanitizeDerivationName;
+
+  toCoordinates =
+    id:
+    let
+      coords = builtins.split ":" id;
+
+      parseVersion =
+        version:
+        let
+          parts = builtins.split ":" version;
+          base = elemAt parts 0;
+        in
+        if length parts >= 2 then
+          let
+            snapshot = elemAt parts 2;
+          in
+          replaceStrings [ "-SNAPSHOT" ] [ "-${snapshot}" ] base
+        else
+          base;
+    in
+    rec {
+      group = elemAt coords 0;
+      module = elemAt coords 2;
+      version = elemAt coords 4;
+      uniqueVersion = parseVersion version;
+    };
+
+  fetchers' = {
+    http = fetchurl;
+    https = fetchurl;
+  } // fetchers;
+
+  fetch =
+    coords: overrides: name:
+    { url, hash }:
+    let
+      scheme = head (builtins.match "([a-z0-9+.-]+)://.*" url);
+      fetch' = getAttr scheme fetchers';
+      artifact = fetch' { inherit url hash; };
+      override = overrides.${name} or lib.id;
+    in
+    override artifact;
+
+  mkModule =
+    id: artifacts:
+    let
+      coords = toCoordinates id;
+      modulePath = "${replaceStrings [ "." ] [ "/" ] coords.group}/${coords.module}/${coords.version}";
+      moduleOverrides = overrides.${id} or { };
+    in
+    stdenv.mkDerivation {
+      pname = sanitizeDerivationName "${coords.group}-${coords.module}";
+      version = coords.uniqueVersion;
+
+      srcs = mapAttrsToList (fetch coords moduleOverrides) artifacts;
+
+      dontPatch = true;
+      dontConfigure = true;
+      dontBuild = true;
+      dontFixup = true;
+      dontInstall = true;
+
+      preUnpack = ''
+        mkdir -p "$out/${modulePath}"
+      '';
+
+      unpackCmd = ''
+        cp "$curSrc" "$out/${modulePath}/$(stripHash "$curSrc")"
+      '';
+
+      sourceRoot = ".";
+
+      preferLocalBuild = true;
+      allowSubstitutes = false;
```

Comment:
Why do we do this? These seem to just be hashed binaries, and it forces people to manually build every dependency instead of download from binary cache

Reply:
This is the equivalent of what `runCommandLocal` or `fetchurl` does; in general, it doesn't make sense to have the substituter perform a fetch, then fetch the mostly-unchanged result from the substituter. You also want to know sooner rather than later that your ultimate source of truth (usually one or more Maven repositories) is providing stable downloads, otherwise you get problems reproducing builds (as you intentionally bypass substituters to do so).

Reply:
This is the equivalent of what `runCommandLocal` or `fetchurl` does; in general, it doesn't make sense to have the substituter perform a fetch, then fetch the mostly-unchanged result from the substituter. You also want to know sooner rather than later that your ultimate source of truth (usually one or more Maven repositories) is providing stable downloads, otherwise you get problems reproducing builds (as you intentionally bypass substituters to do so).

Reply:
For posterity: https://nix.dev/manual/nix/2.22/language/advanced-attributes.html?highlight=allowsubstitutes#adv-attr-allowSubstitutes

I forked this branch and tried it with `allowSubstitues = true;`, but it didn't cause me to download everything from my binary cache anyway, so I no longer feel strongly either way on this :P